server/sonar-server/src/main/java/org/sonar/server/platform/components/ComponentLevel4.java

```diff
@@ -0,0 +1,735 @@
+import org.sonar.server.rule.ws.UpdateAction;
+import org.sonar.server.source.HtmlSourceDecorator;
+import org.sonar.server.source.SourceService;
+import org.sonar.server.source.index.SourceLineIndex;
+import org.sonar.server.source.index.SourceLineIndexDefinition;
+import org.sonar.server.source.index.SourceLineIndexer;
+import org.sonar.server.source.ws.HashAction;
+import org.sonar.server.source.ws.IndexAction;
+import org.sonar.server.source.ws.LinesAction;
+import org.sonar.server.source.ws.RawAction;
+import org.sonar.server.source.ws.ScmAction;
+import org.sonar.server.source.ws.ShowAction;
+import org.sonar.server.source.ws.SourcesWs;
+import org.sonar.server.test.CoverageService;
+import org.sonar.server.test.index.TestIndex;
+import org.sonar.server.test.index.TestIndexDefinition;
+import org.sonar.server.test.index.TestIndexer;
+import org.sonar.server.test.ws.TestsCoveredFilesAction;
+import org.sonar.server.test.ws.TestsListAction;
+import org.sonar.server.test.ws.TestsWs;
+import org.sonar.server.text.MacroInterpreter;
+import org.sonar.server.text.RubyTextService;
+import org.sonar.server.ui.PageDecorations;
+import org.sonar.server.ui.Views;
+import org.sonar.server.ui.ws.ComponentNavigationAction;
+import org.sonar.server.ui.ws.GlobalNavigationAction;
+import org.sonar.server.ui.ws.NavigationWs;
+import org.sonar.server.ui.ws.SettingsNavigationAction;
+import org.sonar.server.updatecenter.ws.UpdateCenterWs;
+import org.sonar.server.user.DefaultUserService;
+import org.sonar.server.user.GroupMembershipFinder;
+import org.sonar.server.user.GroupMembershipService;
+import org.sonar.server.user.NewUserNotifier;
+import org.sonar.server.user.SecurityRealmFactory;
+import org.sonar.server.user.UserUpdater;
+import org.sonar.server.user.index.UserIndex;
+import org.sonar.server.user.index.UserIndexDefinition;
+import org.sonar.server.user.index.UserIndexer;
+import org.sonar.server.user.ws.FavoritesWs;
+import org.sonar.server.user.ws.UserPropertiesWs;
+import org.sonar.server.user.ws.UsersWs;
+import org.sonar.server.util.BooleanTypeValidation;
+import org.sonar.server.util.FloatTypeValidation;
+import org.sonar.server.util.IntegerTypeValidation;
+import org.sonar.server.util.StringListTypeValidation;
+import org.sonar.server.util.StringTypeValidation;
+import org.sonar.server.util.TextTypeValidation;
+import org.sonar.server.util.TypeValidations;
+import org.sonar.server.view.index.ViewIndex;
+import org.sonar.server.view.index.ViewIndexDefinition;
+import org.sonar.server.view.index.ViewIndexer;
+import org.sonar.server.ws.ListingWs;
+import org.sonar.server.ws.WebServiceEngine;
+
+public class ComponentLevel4 extends ComponentLevel {
+
+  private final List<Object> level4AddedComponents;
+
+  public ComponentLevel4(ComponentLevel parent, List<Object> level4AddedComponents) {
+    super("level4", parent);
+    this.level4AddedComponents = level4AddedComponents;
+  }
+
+  @Override
+  public ComponentLevel configure() {
```

Comment:
So this would push the refactoring one step beyond: from one big ServerComponent class, we end up with a tree of components by level. I like this idea.